receiver/kubernetesclusterreceiver/collector.go

```diff
@@ -0,0 +1,213 @@
+	containerKeySpecName = "container.spec.name"
+)
+
+// dataCollector wraps around a metricsStore and a metadaStore exposing
+// methods to perform on the underlying stores
+type dataCollector struct {
+	logger                 *zap.Logger
+	metricsStore           *metricsStore
+	metadataStore          *metadataStore
+	nodeConditionsToReport []string
+}
+
+// newDataCollector returns a dataCollector.
+func newDataCollector(logger *zap.Logger, nodeConditionsToReport []string) *dataCollector {
+	return &dataCollector{
+		logger: logger,
+		metricsStore: &metricsStore{
+			metricsCache: make(map[types.UID][]consumerdata.MetricsData),
+		},
+		metadataStore: &metadataStore{
+			stores: map[string]cache.Store{},
+		},
+		nodeConditionsToReport: nodeConditionsToReport,
+	}
+}
+
+// setupMetadataStore initializes a metadata store for the kubernetes object.
+func (dc *dataCollector) setupMetadataStore(o runtime.Object, store cache.Store) {
+	dc.metadataStore.setupStore(o, store)
+}
+
+func (dc *dataCollector) removeFromMetricsStore(obj interface{}) {
+	key, err := utils.GetUIDForObject(obj.(runtime.Object))
+
+	if err != nil {
+		dc.logger.Error(
+			"failed to remove from metric cache",
+			zap.String("obj", reflect.TypeOf(obj).String()),
+			zap.Error(err),
+		)
+		return
+	}
+
+	dc.metricsStore.remove(key)
+}
+
+func (dc *dataCollector) updateMetricsStore(key types.UID, rms []*resourceMetrics) {
+	dc.metricsStore.update(key, rms)
+}
+
+func (dc *dataCollector) collectMetrics() []consumerdata.MetricsData {
+	return dc.metricsStore.getMetrics()
+}
+
+// syncMetrics updates the metric store with latest metrics from the kubernetes object.
+func (dc *dataCollector) syncMetrics(obj interface{}) {
+	rm := make([]*resourceMetrics, 0)
+
+	switch o := obj.(type) {
+	case *corev1.Pod:
+		rm = getMetricsForPod(o)
+	case *corev1.Node:
+		rm = getMetricsForNode(o, dc.nodeConditionsToReport)
+	case *corev1.Namespace:
+		rm = getMetricsForNamespace(o)
+	case *corev1.ReplicationController:
+		rm = getMetricsForReplicationController(o)
+	case *corev1.ResourceQuota:
+		rm = getMetricsForResourceQuota(o)
+	case *appsv1.Deployment:
+		rm = getMetricsForDeployment(o)
+	case *appsv1.ReplicaSet:
+		rm = getMetricsForReplicaSet(o)
+	case *appsv1.DaemonSet:
+		rm = getMetricsForDaemonSet(o)
+	case *appsv1.StatefulSet:
+		rm = getMetricsForStatefulSet(o)
+	case *batchv1.Job:
+		rm = getMetricsForJob(o)
+	case *batchv1beta1.CronJob:
+		rm = getMetricsForCronJob(o)
+	case *v2beta1.HorizontalPodAutoscaler:
+		rm = getMetricsForHPA(o)
+	case *corev1.Service:
+	default:
+		dc.logger.Warn(
+			"Unsupported k8s resource type",
+			zap.String("obj", reflect.TypeOf(obj).String()),
+		)
+		return
+	}
+
+	if len(rm) == 0 {
+		return
+	}
+
+	key, err := utils.GetUIDForObject(obj.(runtime.Object))
+
+	if err != nil {
+		dc.logger.Error(
+			"failed to update metric cache",
+			zap.String("obj", reflect.TypeOf(obj).String()),
+			zap.Error(err),
+		)
+		return
+	}
```

Comment:
Could this go in `updateMetricsStore`, similar to `removeFromMetricStore`?

Reply:
Moved this logic into `metricStore` like you suggested in the comment below for update and remove methods.